src/expiring_hash_map/mod.rs

```diff
@@ -0,0 +1,200 @@
+    /// expiried items regardless of what state the [`ExpiringHashMap`] is.
+    /// This is a deliberate design decision, we went with it for the following
+    /// reasons:
+    /// 1. Use of `async fn`. One of the benefits of this API is that it relies
+    ///    only on `async fn`s, and doesn't require manual `Future`
+    ///    implementation. While this is not a problem in general, but there is
+    ///    some value with doing it this way. With a switch to `async` across
+    ///    our code base, the idea is that we should completely eliminate manual
+    ///    `Future` implementations and poll fns. This is controversial, but we
+    ///    decided to give it a try.
+    /// 2. We don't know all the use cases, and exposing this kind of API might
+    ///    make more sense, since it allows more flexibility.
+    ///    We were choosing between, effectively, the current "drain"-like API,
+    ///    and the "queue" like API.
+    ///    Current ("drain"-like) API waits on the deadline or returns `None`
+    ///    when there are no more items. Very similar how we [`Vec::drain`] iter
+    ///    works.
+    ///    The "queue"-like API would, pretty much, be simply waiting expired
+    ///    items to appear. In the case of empty [`ExpiringHashMap`], we would
+    ///    wait indefinetly - or until an item is inserted. This would be
+    ///    possible to carry on, for instance, from a sibling branch of a
+    ///    `select` statement, so the borrowing rules won't be a problem here.
+    /// 3. We went over the following alternative signature:
+    ///    ```ignore
+    ///    pub fn next_expired(&mut self) -> Option<impl Future<Outcome = Result<ExpiredItem<K, V>, Error>>> {...}
+    ///    ```
+    ///    This captures the API restrictions a bit better, and should provide
+    ///    less possibilities to misuse the API.
+    ///    We didn't pick this one becuase it's not an `async fn` and we wanted
+    ///    this, see (1) of this list. Furtherore, instead of doing a
+    ///    `select { _ = map.next_expired(), if !map.is_empty() => { ... } }`
+    ///    users would have to do
+    ///    `let exp = map.next_expired(); select { _ = exp.unwrap(), if exp.is_some() => { ... } }`,
+    ///    which is less readable and a bit harder to understand. Although it
+    ///    has a possiblity of a nicer generalization if `select` macro
+    ///    supported a `Some(future)` kind of pattern matching, we decided to go
+    ///    with other solution for now.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use tokio02 as tokio;
+    /// # let mut rt = tokio::runtime::Runtime::new().unwrap();
+    /// # rt.block_on(async {
+    /// use vector::expiring_hash_map::ExpiringHashMap;
+    /// use std::time::Duration;
+    ///
+    /// let mut map: ExpiringHashMap<String, String> = ExpiringHashMap::new();
+    ///
+    /// loop {
+    ///     tokio::select! {
+    ///         // Avoid spin lock!
```

Comment:
I'd change this to say, you need to ensure that this branch is disabled if the map is empty. That is the more important thing to understand. That this select branch should be disabled.

Reply:
Good idea! I changed the comment accordingly.